src/IECoreScene/SceneCache.cpp

```diff
@@ -728,8 +697,232 @@ class SceneCache::ReaderImplementation : public SceneCache::Implementation
+			if( descendentTagsIO )
+			{
+				descendentTagsIO->entryIds( descendentTags );
+			}
+
+			/// provided for backward compatibility...
+			if( tags.empty() && descendentTags.empty() )
+			{
+				ConstIndexedIOPtr tagsIO = m_indexedIO->subdirectory( tagsEntry, IndexedIO::NullIfMissing );
+				if( tagsIO )
+				{
+					tagsIO->entryIds( tags, IndexedIO::File );
+					tagsIO->entryIds( descendentTags );
+				}
+			}
+
+			if( tags.empty() && descendentTags.empty() )
+			{
+				return;
+			}
+
+			m_sets = new CompoundData();
+
+			SceneInterface::Path currentPath;
+			path( currentPath );
+			for( const auto &tag : tags )
+			{
+				getRoot()->appendToSet( tag, currentPath );
+			}
+
+			if( !descendentTags.empty() )
+			{
+				SceneCache::NameList children;
+				childNames( children );
+
+				for( const auto &c : children )
+				{
+					ReaderImplementationPtr childReader = child( c, NullIfMissing );
+					if( childReader )
+					{
+						childReader->convertTagsToSets();
+					}
+				}
+			}
+		}
+
 	private :
 
+		void appendToSet( const Name &setName, const SceneInterface::Path &path )
+		{
+			auto &writable = m_sets->writable();
+			auto it = writable.find( setName );
+			if( it == writable.end() )
+			{
+				it = writable.insert( std::make_pair( setName, new IECore::PathMatcherData() ) ).first;
+			}
+			PathMatcherDataPtr pathMatcher = runTimeCast<PathMatcherData>( it->second );
+			pathMatcher->writable().addPath( path );
+		}
+
+		PathMatcherDataPtr readLocalSet( const Name &name ) const
+		{
+			IECore::CompoundDataBasePtr compoundDataBase = getSetCompound();
+
+			if( !compoundDataBase )
+			{
+				return new IECore::PathMatcherData();
+			}
+
+			CompoundDataPtr compoundData = new CompoundData( compoundDataBase->readable() );
+
+			if( IECore::PathMatcherDataPtr foundPathMatcher = compoundData->member<PathMatcherData>( name ) )
+			{
+				return foundPathMatcher;
+			}
+
+			return new IECore::PathMatcherData();
+		}
+
+		NameList readChildSetNames() const
+		{
+			IndexedIO::EntryIDList childSetNames;
+			IndexedIOPtr childSets = m_indexedIO->subdirectory( childSetsEntry, IECore::IndexedIO::NullIfMissing );
+			if( childSets )
+			{
+				childSets->entryIds( childSetNames );
+			}
+			else
+			{
+				IndexedIOPtr descenentTags = m_indexedIO->subdirectory( descendentTagsEntry, IECore::IndexedIO::NullIfMissing);
+				if ( descenentTags )
+				{
+					descenentTags->entryIds( childSetNames );
+				}
+			}
+
+			return childSetNames;
+		}
+
+		IECore::CompoundDataBasePtr getSetCompound() const
+		{
+			if( m_sets )
+			{
+				return runTimeCast<CompoundDataBase>( m_sets );
+			}
+
+			// todo: every time we read a set we load from the indexed IO ?
+
+			IECore::PathMatcherDataPtr result = new IECore::PathMatcherData();
+
+			IndexedIOPtr sets = m_indexedIO->subdirectory( setsEntry, IECore::IndexedIO::NullIfMissing );
+			if( !sets )
+			{
+				return nullptr;
+			}
+
+			ObjectPtr ptr = IECore::Object::load( sets, localSetsEntry );
```

Comment:
This looks suspiciously like we load all sets at once? If so, I don't think that is acceptable from a performance standpoint.

Reply:
> This looks suspiciously like we load all sets at once? If so, I don't think that is acceptable from a performance standpoint.

We used to load tags for the entire heirachy up front, by virtue of them being defined as directories  before didn't we? I'll separate this in to a directory per set. 

Regarding performance, I think we need some specific performance regression tests which I'll add once the design is on a better footing.

Reply:
> We used to load tags for the entire heirachy up front, by virtue of them being defined as directories before didn't we? I'll separate this in to a directory per set.

We did, but I think we can do better here. Also, we need to do better so that we can load the names of the sets without loading the sets themselves. Directory per set sounds like the answer...